### ISSUE TITLE
npm is crashing on mips64le

### ISSUE THREAD

**Issue by vanloswang** [npm/npm#15156]
#### I'm opening this issue because:

  - [x] npm is crashing.

#### What's going wrong?
compile and install nodejs 6.9.1 with npm 3.10.8 on mips64le platform, run 'npm -v', then it hangs up and nerver exit, and it's cpu usage reaches almost 100%.

#### How can the CLI team reproduce the problem?

1、get source code from http://dl.fedoraproject.org/pub/fedora/linux/releases/25/Everything/source/tree/Packages/n/nodejs-6.9.1-1.fc25.src.rpm
2、compile and install on mips64le platform
3、run ‘npm -v’

### supporting information:

I use strace to trace the 'npm -v' cmd and get the following information:
```bash
stat("/usr/lib/node_modules/npm/lib/utils/umask", 0xffffc755a0) = -1 ENOENT (No such file or directory)
stat("/usr/lib/node_modules/npm/lib/utils/umask.js", {st_mode=S_IFREG|0644, st_size=343, ...}) = 0
lstat("/usr", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/lib", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/lib/utils", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/lib/utils/umask.js", {st_mode=S_IFREG|0644, st_size=343, ...}) = 0
open("/usr/lib/node_modules/npm/lib/utils/umask.js", O_RDONLY|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFREG|0644, st_size=343, ...}) = 0
read(12, "var umask = require('umask')\nvar"..., 343) = 343
close(12)                               = 0
cacheflush(0x3a682660, 0x3a4, 0x1)      = 0
cacheflush(0x3a682a80, 0x144, 0x1)      = 0
stat("/usr/lib/node_modules/npm/lib/utils/node_modules", 0xffffc75300) = -1 ENOENT (No such file or directory)
stat("/usr/lib/node_modules/npm/node_modules/umask", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
open("/usr/lib/node_modules/npm/node_modules/umask/package.json", O_RDONLY|O_CLOEXEC) = 12
pread64(12, "{\n  \"name\": \"umask\",\n  \"version\""..., 32768, 0) = 1287
pread64(12, "", 32768, 1287)            = 0
close(12)                               = 0
stat("/usr/lib/node_modules/npm/node_modules/umask/index.js", {st_mode=S_IFREG|0644, st_size=2008, ...}) = 0
lstat("/usr", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib", {st_mode=S_IFDIR|0555, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/node_modules", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/node_modules/umask", {st_mode=S_IFLNK|0777, st_size=29, ...}) = 0
stat("/usr/lib/node_modules/npm/node_modules/umask", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
readlink("/usr/lib/node_modules/npm/node_modules/umask", "../node_modules.bundled/umask", 4096) = 29
lstat("/usr/lib/node_modules/npm/node_modules.bundled", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/node_modules.bundled/umask", {st_mode=S_IFDIR|0755, st_size=4096, ...}) = 0
lstat("/usr/lib/node_modules/npm/node_modules.bundled/umask/index.js", {st_mode=S_IFREG|0644, st_size=2008, ...}) = 0
open("/usr/lib/node_modules/npm/node_modules.bundled/umask/index.js", O_RDONLY|O_CLOEXEC) = 12
fstat(12, {st_mode=S_IFREG|0644, st_size=2008, ...}) = 0
read(12, "'use strict';\n\nvar util = requir"..., 2008) = 2008
close(12)                               = 0
cacheflush(0x3a682c40, 0x4f4, 0x1)      = 0
cacheflush(0x3a6831a0, 0x144, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 175503426}) = 0
cacheflush(0x3a683360, 0x320, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 176049567}) = 0
cacheflush(0x3a6834b4, 0x10, 0x1)       = 0
cacheflush(0x3a683494, 0x8, 0x1)        = 0
cacheflush(0x3a6836e0, 0x28, 0x1)       = 0
cacheflush(0x3a6836e0, 0x10, 0x1)       = 0
cacheflush(0x3a68352c, 0x10, 0x1)       = 0
cacheflush(0x3a65be58, 0x10, 0x1)       = 0
cacheflush(0x3a65bed0, 0x10, 0x1)       = 0
cacheflush(0x3a65bf08, 0x10, 0x1)       = 0
cacheflush(0x3a65bf28, 0x10, 0x1)       = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 178135981}) = 0
cacheflush(0x3a683780, 0x654, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 178981635}) = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 179193557}) = 0
cacheflush(0x3a683e40, 0x134, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 179551032}) = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 179662296}) = 0
cacheflush(0x3a683fe0, 0x8c4, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 180369053}) = 0
cacheflush(0x3a684044, 0x10, 0x1)       = 0
cacheflush(0x3a68449c, 0x10, 0x1)       = 0
cacheflush(0x3a684580, 0x10, 0x1)       = 0
cacheflush(0x3a684664, 0x10, 0x1)       = 0
cacheflush(0x3a684920, 0x528, 0x1)      = 0
cacheflush(0x3a684740, 0x10, 0x1)       = 0
cacheflush(0x3a6838a4, 0x10, 0x1)       = 0
cacheflush(0x3a683b6c, 0x10, 0x1)       = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 181853292}) = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 181964777}) = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 182078535}) = 0
cacheflush(0x3a684ec0, 0x134, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 182444898}) = 0
cacheflush(0x3a65c284, 0x10, 0x1)       = 0
getuid()                                = 0
cacheflush(0x3a65c3dc, 0x10, 0x1)       = 0
cacheflush(0x3a65c690, 0x10, 0x1)       = 0
cacheflush(0x3a65c78c, 0x10, 0x1)       = 0
cacheflush(0x3a65c7cc, 0x10, 0x1)       = 0
cacheflush(0x3a65c8b0, 0x10, 0x1)       = 0
cacheflush(0x3a685060, 0x90, 0x1)       = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 185307419}) = 0
cacheflush(0x3a685160, 0x614, 0x1)      = 0
clock_gettime(CLOCK_MONOTONIC, {164162, 186200359}) = 0
socket(PF_NETLINK, SOCK_RAW, NETLINK_ROUTE) = 12
bind(12, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 0
getsockname(12, {sa_family=AF_NETLINK, pid=28984, groups=00000000}, [12]) = 0
gettimeofday({1481091773, 100123}, NULL) = 0
sendto(12, "\24\0\0\0\22\0\1\3\275\252GX\0\0\0\0\0\0\0\0", 20, 0, {sa_family=AF_NETLINK, pid=0, groups=00000000}, 12) = 20
recvmsg(12, {msg_name(12)={sa_family=AF_NETLINK, pid=0, groups=00000000}, msg_iov(1)=[{"\214\4\0\0\20\0\2\0\275\252GX8q\0\0\0\0\4\3\1\0\0\0I\0\1\0\0\0\0\0"..., 4096}], msg_controllen=0, msg_flags=MSG_TRUNC}, 0) = 4096
close(12)                               = 0
--- SIGBUS {si_signo=SIGBUS, si_code=SI_KERNEL, si_addr=0} ---
read(3, "*", 1)                         = 1
write(7, "8\231\7!\1\0\0\0\n\0\0\0\0\0\0\0", 16) = 16
write(4, "*", 1)                        = 1
rt_sigreturn()                          = -1 ERRNO_18446744073678225424 (Unknown error -31326192)
--- SIGBUS {si_signo=SIGBUS, si_code=SI_KERNEL, si_addr=0} ---
read(3, "*", 1)                         = 1
write(7, "8\231\7!\1\0\0\0\n\0\0\0\0\0\0\0", 16) = 16
write(4, "*", 1)                        = 1
rt_sigreturn()                          = -1 ERRNO_18446744073678225424 (Unknown error -31326192)
--- SIGBUS {si_signo=SIGBUS, si_code=SI_KERNEL, si_addr=0} ---
read(3, "*", 1)                         = 1
write(7, "8\231\7!\1\0\0\0\n\0\0\0\0\0\0\0", 16) = 16
write(4, "*", 1)                        = 1
rt_sigreturn()                          = -1 ERRNO_18446744073678225424 (Unknown error -31326192)
--- SIGBUS {si_signo=SIGBUS, si_code=SI_KERNEL, si_addr=0} ---
read(3, "*", 1)                         = 1
write(7, "8\231\7!\1\0\0\0\n\0\0\0\0\0\0\0", 16) = 16
write(4, "*", 1)                        = 1
rt_sigreturn()                          = -1 ERRNO_18446744073678225424 (Unknown error -31326192)
--- SIGBUS {si_signo=SIGBUS, si_code=SI_KERNEL, si_addr=0} ---
read(3, "*", 1)                         = 1
write(7, "8\231\7!\1\0\0\0\n\0\0\0\0\0\0\0", 16) = 16
write(4, "*", 1)                        = 1
rt_sigreturn()                          = -1 ERRNO_18446744073678225424 (Unknown error -31326192)
```
it is loop and says the same error here. Could someone give me some suggestions to debug this issue out?

**Comment by legodude17** [comment]
This may be realated: nodejs/node#10122.

**Comment by bnoordhuis** [comment]
I think the repeated SIGBUS errors are caused by npm's signal-exit module installing a handler for that signal.  That's a very wrong thing to do if you aren't going to actually handle the bus error - and since you can't from JS land there is no point in installing a listener for it.

signal-exit should fix that and it should fix that for other signals it tries to intercept as well.  There is probably is a null pointer dereference somewhere (si_code=SI_KERNEL and si_addr=0 practically guarantees it) but the signal handler hides the real bug.

**Comment by legodude17** [comment]
So that means this isn't a npm issue, right?

**Comment by bnoordhuis** [comment]
The null pointer dereference?  Probably not.  The SIGBUS loop?  npm bug.

**Comment by iarna** [comment]
This was fixed with https://github.com/tapjs/signal-exit/commit/5b105fb9e3f1c08c781141fd15bd0a79ad0293c8 and was landed `npm@4.0.0` with cb20c73.

To update on a platform where executing any npm command causes this error you'll have to use the old shell script installer to upgrade:

```
curl http://npmjs.org/install.sh | sudo sh
```
